### PR TITLE
Split wasm engines into external/internal headers

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by axic** [comment]
@chfast there is probably a better way to do this. I'll let you take care.

**Comment by axic** [comment]
> Move the contents of X-internal.h to X.cpp.

Isn't that against the main goal of having a clear separation of class definition (gives a good overview what the members/functions are) and implementation?

File: src/binaryen.cpp [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include <wasm-validator.h>
 
 #include "binaryen.h"
+#include "binaryen-internal.h"
```
Comment (chfast) [inline src/binaryen.cpp]: It's best when headers matching a .cpp file are included as first. Then you avoid situation when the headers misses any include.

**Comment by axic** [comment]
@chfast did the split for wavm. Can you check if this is what you've meant?

File: src/binaryen.cpp [hunk 2]
```diff
@@ -434,8 +467,10 @@ namespace hera {
     heraAssert(false, string("Unsupported import called: ") + import->module.str + "::" + import->base.str + " (" + to_string(arguments.size()) + "arguments)");
   }
 
+namespace {
+
 // NOTE: This should be caught during deployment time by the Sentinel.
-void BinaryenEngine::validate_contract(Module & module)
+void validate_contract(Module & module)
```
Comment (axic) [inline src/binaryen.cpp]: @chfast Which is better, this in an anonymous namespace or an empty forward declaration of `wasm::Module` in `binaryen.h` - see the last commit for that.

**Comment by axic** [comment]
> but this will mess up the other WAVM PR.

Shouldn't really affect it, once it is rebased. Currently it duplicates some of the commits already merged on a different PR.

File: src/binaryen.cpp (hunk 2 above)
Comment (chfast) [reply src/binaryen.cpp]: If possible, hide it from the header.

Comment (axic) [reply src/binaryen.cpp]: It can stay in an anonymous namespace.

**Comment by axic** [comment]
@chfast final version ^^

**Pull request merged**